free45/toplevels/bsg_ac_black_parrot_single_core_rc3/tcl/hard/gf_14/placement_constraints.tcl

```diff
@@ -0,0 +1,301 @@
+  -anchor_corner br \
+  -offset [list 0 0]
+
+#####################################
+### D CACHE TAG
+###
+
+set dcache_tag_mems [get_cells -design bp_tile_node -hier -filter "ref_name=~gf14_* && full_name=~*/dcache/tag_mem*"]
+set dcache_tag_ma [create_macro_array \
+  -num_rows 1 \
+  -num_cols 2 \
+  -align bottom \
+  -horizontal_channel_height [expr 2*$keepout_margin_y] \
+  -vertical_channel_width [expr 2*$keepout_margin_x] \
+  -orientation N \
+  $dcache_tag_mems]
+
+create_keepout_margin -type hard -outer $keepout_margins $dcache_tag_mems
+
+set dcache_tag_margin 0
+set_macro_relative_location \
+  -target_object $dcache_tag_ma \
+  -target_corner br \
+  -target_orientation R0 \
+  -anchor_object $dcache_data_ma \
+  -anchor_corner tr \
+  -offset [list -$dcache_tag_margin 0]
+
+#####################################
+### D CACHE STAT
+###
+
+#set dcache_stat_mem [get_cells -design bp_tile_node -hier -filter "ref_name=~gf14_* && full_name=~*/dcache/stat_mem*"]
+#set dcache_stat_margin 0
+#set_macro_relative_location \
+#  -target_object $dcache_stat_mem \
+#  -target_corner br \
+#  -target_orientation R0 \
+#  -anchor_object $dcache_tag_ma \
+#  -anchor_corner bl \
+#  -offset [list -$dcache_stat_margin $keepout_margin_y]
+#
+#create_keepout_margin -type hard -outer $keepout_margins $dcache_stat_mem
+
+#####################################
+### CCE Directory
+###
+
+set directory_mems [get_cells -hier -filter "ref_name=~gf14_* && full_name=~*bp_cce/directory/directory/*"]
+set directory_mem_height [get_attribute -objects [index_collection $directory_mems 0] -name height]
+set directory_mem_width [get_attribute -objects [index_collection $directory_mems 0] -name width]
+set directory_ma [create_macro_array \
+  -num_rows 1 \
+  -num_cols 8 \
+  -align bottom \
+  -horizontal_channel_height [expr 2*$keepout_margin_y] \
+  -vertical_channel_width [expr 2*$keepout_margin_x] \
+  -orientation [list N N N N N N N N] \
+  $directory_mems]
+
+# Should put this in the middle by relative location
+set_macro_relative_location \
+  -target_object $directory_ma \
+  -target_corner tl \
+  -target_orientation R0 \
+  -anchor_corner tl \
+  -offset [list [expr $tile_width/2 - 4*$directory_mem_width] 0]
+
+create_keepout_margin -type hard -outer $keepout_margins $directory_mems
+
+#####################################
+### CCE Instance
+###
+
+set cce_instr_ram [get_cells -design bp_tile_node -hier -filter "ref_name=~gf14_* && full_name=~*/bp_cce/*inst_ram*"]
+set cce_instr_width [get_attribute -objects $cce_instr_ram -name width]
+set cce_instr_height [get_attribute -objects $cce_instr_ram -name height]
+set_macro_relative_location \
+  -target_object $cce_instr_ram \
+  -target_corner tr \
+  -target_orientation R0 \
+  -anchor_object $directory_ma \
+  -anchor_corner tl \
+  -offset [list -$keepout_margin_x -$keepout_margin_y]
+
+create_keepout_margin -type hard -outer $keepout_margins $cce_instr_ram
+
+#####################################
```

Comment:
BTB is hardened, no?

Reply:
I haven't touched the post synth scripts yet. I just carried these over from the quad_core_rc3. When I get to work on them, I will update them to reflect your recent rc5 changes.

Reply:
okay!